### PR TITLE
Preprocess (draft request)

### PR DESCRIPTION
**Pull request by hkershaw-brown**
Draft review so I can go through the changes and collect todos.

### CODE REVIEW

File: assimilation_code/modules/observations/atmosphere_quantities.f90 [hunk 1]
```diff
@@ -0,0 +1,197 @@
+! DART software - Copyright UCAR. This open source software is provided
```
Comment (hkershaw-brown) [inline assimilation_code/modules/observations/atmosphere_quantities.f90:1]: Are the units documented anywhere?

File: developer_tests/utilities/parse_args_test.f90 [hunk 2]
```diff
@@ -0,0 +1,70 @@
+! DART software - Copyright UCAR. This open source software is provided
```
Comment (hkershaw-brown) [inline developer_tests/utilities/parse_args_test.f90:1]: where is the test here?

File: developer_tests/utilities/work/mkmf_parse_args_test [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+# by UCAR, "as is", without charge, subject to all terms of use at
+# http://www.image.ucar.edu/DAReS/DART/DART_download
+#
+# $Id: mkmf_parse_args_test 11946 2017-09-21 22:10:54Z nancy@ucar.edu $
+
+../../../build_templates/mkmf -p parse_args_test -t ../../../build_templates/mkmf.template   \
+        -a "../../.." path_names_parse_args_test
+
+# <next few lines under version control, do not edit>
+# $URL: https://svn-dares-dart.cgd.ucar.edu/DART/branches/preprocess/developer_tests/utilities/work/mkmf_parse_args_test $
```
Comment (hkershaw-brown) [inline developer_tests/utilities/work/mkmf_parse_args_test:13]: end of file subversion junk to remove

File: docs/pages/Getting_Started.md [hunk 4]
```diff
@@ -2332,24 +2332,24 @@ observation-specific code for reading, writing, or initializing the
 observation, then the entire contents of the new file is:
 
 ~~~
-! BEGIN DART PREPROCESS KIND LIST
+! BEGIN DART PREPROCESS QUANTITY LIST
```
Comment (hkershaw-brown) [inline docs/pages/Getting_Started.md:2335]: check 'quantity' vs 'kind' in this doc.

Comment (hkershaw-brown) [reply docs/pages/Getting_Started.md:2335]: No longer correct

````
Please see the list of available observation quantities in
`assimilation_code/modules/observations/DEFAULT_obs_kind_mod.f90` to determine	`assimilation_code/modules/observations/DEFAULT_obs_kind_mod.f90`
````

File: models/9var/work/input.nml [hunk 5]
```diff
@@ -144,9 +144,10 @@
 &preprocess_nml
     input_obs_def_mod_file = '../../../observations/forward_operators/DEFAULT_obs_def_mod.F90',
```
Comment (hkershaw-brown) [inline models/9var/work/input.nml:145]: Does anyone ever change 
input_obs_def
output_obs_def?

File: assimilation_code/programs/fill_inflation_restart/fill_inflation_restart.f90 [hunk 6]
```diff
@@ -327,3 +327,9 @@ end subroutine finalize_modules_used
 
 end program fill_inflation_restart
 
+! <next few lines under version control, do not edit>
```
Comment (hkershaw-brown) [inline assimilation_code/programs/fill_inflation_restart/fill_inflation_restart.f90:330]: remove svn junk

File: assimilation_code/programs/fill_inflation_restart/fill_inflation_restart.html [hunk 7]
```diff
@@ -32,6 +32,7 @@ <H2>Overview</H2>
 
 <P>
 Utility program to create inflation restart files with constant values.
+
```
Comment (hkershaw-brown) [inline assimilation_code/programs/fill_inflation_restart/fill_inflation_restart.html:35]: more random whitespace difference

File: assimilation_code/programs/obs_utils/create_obs_grid.f90 [hunk 8]
```diff
@@ -14,7 +14,7 @@
 
 program create_obs_grid
 
-use    utilities_mod, only : register_module, open_file, close_file, &
+use    utilities_mod, only : register_module, &
```
Comment (hkershaw-brown) [inline assimilation_code/programs/obs_utils/create_obs_grid.f90:17]: this does not seem to be related to the new preprocess method

File: assimilation_code/programs/obs_utils/obs_remove_dups.f90 [hunk 9]
```diff
@@ -4,6 +4,8 @@
 !
 ! DART $Id$
```
Comment (hkershaw-brown) [inline assimilation_code/programs/obs_utils/obs_remove_dups.f90:6]: not related to the new preprocess method

File: developer_tests/utilities/parse_args_test.f90 (hunk 2 above)
Comment (nancycollins) [reply developer_tests/utilities/parse_args_test.f90:1]: in the work dir there is a parse_args_test.in - it needs to run parse_args_test < parse_args_test.in

File: docs/pages/Getting_Started.md (hunk 4 above)
Comment (nancycollins) [reply docs/pages/Getting_Started.md:2335]: finding and fixing old lines in code will require grep rather than execution-time tests, because right now preprocess accepts both these lines as synonyms for backwards compatibility.  but we definitely should be documenting the QUANTITY version.

File: models/9var/work/input.nml (hunk 5 above)
Comment (nancycollins) [reply models/9var/work/input.nml:145]: not that i'm aware of, but if someone wanted to test a different version of the template they could use this facility.

File: assimilation_code/modules/observations/atmosphere_quantities.f90 (hunk 1 above)
Comment (nancycollins) [reply assimilation_code/modules/observations/atmosphere_quantities.f90:1]: nope.  huge can of worms.  run away.   (seriously, i think the 'udunits' package that netcdf and other programs use is probably a good starting point for names of units.)

**Comment by hkershaw-brown** [comment]
Just FYI for those with email notices turned up to 11.  This is a draft pull request, I'm using it to go though 211 files to see what needs to be in the actual pull request.  No need to review this unit the actual pull request.

File: assimilation_code/modules/observations/atmosphere_quantities.f90 (hunk 1 above)
Comment (hkershaw-brown) [reply assimilation_code/modules/observations/atmosphere_quantities.f90:1]: Todo: what should go in the documentation about the units column

File: assimilation_code/modules/observations/ocean_quantities.f90 [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+! 
+!
+! ! fixme - this is hardcoded in obs_diag and shouldn't be
+!     QTY_U_WIND_COMPONENT            "none"  MISSING_R8    MISSING_R8
+!     QTY_V_WIND_COMPONENT            "none"  MISSING_R8    MISSING_R8
+! 
+! END DART PREPROCESS QUANTITY DEFINITIONS
+
+
+! <next few lines under version control, do not edit>
```
Comment (hkershaw-brown) [inline assimilation_code/modules/observations/ocean_quantities.f90:38]: svn junk

File: assimilation_code/modules/observations/atmosphere_quantities.f90 [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+!     QTY_HYDROPHILIC_ORGANIC_MATTER    "none"   MISSING_R8   MISSING_R8   ! Hydrophilic organic matter CAMS aerosol (OMAT)
+!     QTY_CLOUDWATER_DE                 "none"   MISSING_R8   MISSING_R8   ! Cloud liquid water effective diameter (microns)
+!     QTY_CLOUD_ICE_DE                  "none"   MISSING_R8   MISSING_R8   ! Cloud ice effective diameter (microns)
+!     QTY_COLUMN_CLOUD_FRAC             "none"   MISSING_R8   MISSING_R8   ! Simple cloud fraction (0-1)
+!     QTY_CLOUD_TOP_PRESSURE            "none"   MISSING_R8   MISSING_R8
+! 
+! END DART PREPROCESS QUANTITY DEFINITIONS
+
+
+! <next few lines under version control, do not edit>
```
Comment (hkershaw-brown) [inline assimilation_code/modules/observations/atmosphere_quantities.f90:193]: svn junk

File: assimilation_code/modules/observations/oned_quantities.f90 [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+!  QTY_1D_INTEGRAL         none      ! compute value with an integral 
+!  QTY_STATE_VAR_POWER     none      ! raising a state value to a power
+!  QTY_LARGE_SCALE_STATE   none      ! state varies with large time/space scale
+!  QTY_SMALL_SCALE_STATE   none      ! state varies with small time/space scale
+!  QTY_1D_PARAMETER        none   
+! 
+! END DART PREPROCESS QUANTITY DEFINITIONS
+
+
+! <next few lines under version control, do not edit>
```
Comment (hkershaw-brown) [inline assimilation_code/modules/observations/oned_quantities.f90:24]: svn junk

File: assimilation_code/modules/observations/seaice_quantities.f90 [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+!   QTY_SEAICE_FY                       "none"
+!   QTY_SEAICE_AGREG_FY                 "none"
+!   QTY_SEAICE_AGREG_SURFACETEMP        "none"
+!   QTY_SEAICE_AGREG_FREEBOARD          "none" 
+!   QTY_SEAICE_CATEGORY                 "none"
+! 
+! END DART PREPROCESS QUANTITY DEFINITIONS
+
+
+! <next few lines under version control, do not edit>
```
Comment (hkershaw-brown) [inline assimilation_code/modules/observations/seaice_quantities.f90:69]: svn junk

File: assimilation_code/modules/observations/space_quantities.f90 [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+!   QTY_VELOCITY_VERTICAL_N4S           "none"   MISSING_R8   MISSING_R8
+!   QTY_VELOCITY_VERTICAL_NO            "none"   MISSING_R8   MISSING_R8
+!   QTY_GND_GPS_VTEC                    "none"   MISSING_R8   MISSING_R8
+!   QTY_DENSITY_ION_OP                  "none"   MISSING_R8   MISSING_R8
+!  
+! 
+! END DART PREPROCESS QUANTITY DEFINITIONS
+
+
+! <next few lines under version control, do not edit>
```
Comment (hkershaw-brown) [inline assimilation_code/modules/observations/space_quantities.f90:123]: svn junk

File: assimilation_code/programs/preprocess/preprocess.html [hunk 15]
```diff
@@ -280,12 +280,12 @@ <H2>ERROR CODES and CONDITIONS</H2>
 </TR>
 
 <TR><!-- routine --><TD VALIGN=top>preprocess</TD>
-    <!-- message --><TD VALIGN=top>file _____ does NOT contain ! BEGIN DART PREPROCESS KIND LIST</TD>
+    <!-- message --><TD VALIGN=top>file _____ does NOT contain ! BEGIN DART PREPROCESS QUANTITY LIST</TD>
```
Comment (hkershaw-brown) [inline assimilation_code/programs/preprocess/preprocess.html:283]: Todo: update preprocess.html

File: assimilation_code/programs/preprocess/preprocess.nml [hunk 16]
```diff
@@ -4,5 +4,6 @@
    output_obs_def_mod_file = '../../../observations/forward_operators/obs_def_mod.f90',
    input_obs_kind_mod_file = '../../../assimilation_code/modules/observations/DEFAULT_obs_kind_mod.F90',
```
Comment (hkershaw-brown) [inline assimilation_code/programs/preprocess/preprocess.nml:5]: needs updating kind-> qty

File: developer_tests/utilities/parse_args_test.f90 [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+call error_handler(E_MSG,'parse_args_test','Finished successfully.',source,revision,revdate)
+call finalize_utilities()
+
+! end of main code
+
+!----------------------------------------------------------------------
+
+end program
+
+! <next few lines under version control, do not edit>
```
Comment (hkershaw-brown) [inline developer_tests/utilities/parse_args_test.f90:66]: svn junk

File: developer_tests/utilities/parse_args_test.f90 [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+
+use utilities_mod,     only : initialize_utilities, finalize_utilities, &
+                              register_module, error_handler, &
+                              E_ERR, E_MSG
+use parse_args_mod,    only : get_args_from_string
+
+implicit none
+
+! version controlled file description for error handling, do not edit
+character(len=*), parameter :: source   = &
```
Comment (hkershaw-brown) [inline developer_tests/utilities/parse_args_test.f90:24]: svn junk

File: developer_tests/utilities/work/mkmf_parse_args_test [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+# DART software - Copyright UCAR. This open source software is provided
+# by UCAR, "as is", without charge, subject to all terms of use at
+# http://www.image.ucar.edu/DAReS/DART/DART_download
+#
+# $Id: mkmf_parse_args_test 11946 2017-09-21 22:10:54Z nancy@ucar.edu $
+
+../../../build_templates/mkmf -p parse_args_test -t ../../../build_templates/mkmf.template   \
+        -a "../../.." path_names_parse_args_test
+
+# <next few lines under version control, do not edit>
```
Comment (hkershaw-brown) [inline developer_tests/utilities/work/mkmf_parse_args_test:12]: svn junk

File: docs/pages/Getting_Started.md (hunk 4 above)
Comment (hkershaw-brown) [reply docs/pages/Getting_Started.md:2335]: documentation change

**Comment by hkershaw-brown** [comment]
Requires input.nml changes for all obs converters. 
Plus models/noah/README.rst

**Review by hkershaw-brown — COMMENTED** [review]
closing this. 
I don't think the half changing kind -> qty is a good idea. 

Many namelist changes that look like this:

   input_obs_**qty**\_mod_file = '../../../assimilation_code/modules/observations/DEFAULT_obs_**kind**\_mod.F90'
